### PR TITLE
Add back the --enable-fast-restart option

### PR DESCRIPTION
**Pull request by FrancoisAi**
Restore the mmap_file function in mtcp_restart that load directly
  into the memory the saved checkpoint file.
  This behavior requires one to avoid using the gzip compression mechanism
  as it is using a pipe to read the gzip-ed file.
  To avoid this missbehavior, the --enable-fast-restart also silently
  forces the dmtcp_launch binary to use --no-gzip option.
  This bug was introduced with the commit id f01c300996fcde04d1a4f6c2ac6d3f39e8a32bc4 
  that went into the version 2.1.1.

### CODE REVIEW

File: src/plugin/Makefile.in [hunk 1]
```diff
@@ -353,7 +354,7 @@ HAS_PYTHON = @HAS_PYTHON@
 HAS_READLINE = @HAS_READLINE@
 HAS_SCREEN = @HAS_SCREEN@
 HAS_SCRIPT = @HAS_SCRIPT@
-HAS_SSH = @HAS_SSH@
+HAS_SSH_LOCALHOST = @HAS_SSH_LOCALHOST@
```
Comment (rohgarg) [inline src/plugin/Makefile.in]: This change seems unrelated to this PR.

File: src/mtcp/mtcp_restart.c [hunk 2]
```diff
@@ -861,6 +862,14 @@ static int read_one_memory_area(int fd)
     }
   }
 
+#ifdef FAST_RST_VIA_MMAP
+    //int should_mmap_ckpt_image = 1;
```
Comment (rohgarg) [inline src/mtcp/mtcp_restart.c]: This commented out variable seems irrelevant.

File: src/mtcp/mtcp_restart.c [hunk 3]
```diff
@@ -861,6 +862,14 @@ static int read_one_memory_area(int fd)
...
+    else if ( 1 /* should_mmap_ckpt_image */ && (area.flags & MAP_ANONYMOUS)) {
```
Comment (rohgarg) [inline src/mtcp/mtcp_restart.c]: So, we only mmap in ANONYMOUS regions. Why? Could you please explain that in a comment?

File: src/mtcp/mtcp_restart.c [hunk 4]
```diff
@@ -861,6 +862,14 @@ static int read_one_memory_area(int fd)
...
+      mmapfile (fd, area.addr, area.size, area.prot | PROT_WRITE,
+                area.flags & ~MAP_ANONYMOUS);
```
Comment (rohgarg) [inline src/mtcp/mtcp_restart.c]: Why do we give the pages write permissions? Again, a comment would be useful.

File: src/plugin/Makefile.in (hunk 1 above)
Comment (FrancoisAi) [inline src/plugin/Makefile.in]: This has been changed automatically by the latest configure.ac file. @gc00 had updated configure.ac but not this file in earlier pull request, because this variable is not used in source plugin, it's an artifact of automake.

File: src/mtcp/mtcp_restart.c (hunk 4 above)
Comment (FrancoisAi) [inline src/mtcp/mtcp_restart.c]: It will also be explained in the next PR update with comments.

**Comment by FrancoisAi** [comment]
@rohgarg I added the comment regarding previous questions.

**Comment by gc00** [comment]
@rohgarg Ping.  Is this now good to go? Thanks.

File: src/mtcp/mtcp_restart.c [hunk 5]
```diff
@@ -861,6 +862,20 @@ static int read_one_memory_area(int fd)
...
+#ifdef FAST_RST_VIA_MMAP
+    /* CASE MAP_ANONYMOUS with FAST_RST enabled
+     * We only want to do this in the MAP_ANONYMOUS case, since we don't want
+     *   any writes to RAM to be reflected back into the underlying file.
+     * Note that in order to map from a file (ckpt image), we must turn off
+     *   anonymous (~MAP_ANONYMOUS).  It's okay, since the fd
+     *   should have been opened with read permission, only.
+     */
+    else if (area.flags & MAP_ANONYMOUS) {
+      mmapfile (fd, area.addr, area.size, area.prot | PROT_WRITE,
```
Comment (rohgarg) [inline src/mtcp/mtcp_restart.c]: The reason why we add write permissions is still unclear to me. Do we remove these permissions later? If not, it seems like this conflicts with the open mode of the file which, as the comment above suggests, should have been read-only.

Comment (gc00) [inline src/mtcp/mtcp_restart.c]: @rohgarg, In fast restart, we want to mmap from the checkpoint image file to RAM (virtual memory).  When we'e done, we want the RAM to have read-write permission if it had read-write permission before checkpointing.  So in this case, we will never turn off the write permission for RAM.

On the other hand, we do not want the writes to RAM to be reflected all the way back into modifying the checkpoint image file itself.  So, ideally, we would choose MAP_ANONYMOUS mode.  But this is not possible, since that mode would not allow us to read from the checkpoint image file.  So, we must _not_ used MAP_ANONYMOUS.  So, to ensure that writes to RAM are not reflected back into modifying the checkpoint image file, we don't give write permission to the fd that points to the checkpoint image file.

So, in summary, the fd pointing to the checkpoint image file never has write permission.  On the other hand, the PROT_WRITE permission for the RAM is always turned on, since we always want to be able to write back to RAM and modify it.

We can add another statement in the comment if you like.  Suppose we wrote, "PROT_WRITE gives us permission to modify the RAM, while the lack of write permission for the fd ensures that we never write back to the checkpoint image file itself."

Would that comment make it clearer?

**Comment by gc00** [comment]
@rohgarg, Ping.  :-)  I think the ball is in your court on this one.  Or maybe I'm forgetting something that we discussed face-to-face.  If it helps, let's talk on the phone or face-to-face.

File: src/mtcp/mtcp_restart.c (hunk 5 above)
Comment (rohgarg) [inline src/mtcp/mtcp_restart.c]: @gc00: I tested locally with `dmtcp1` after removing the `PROT_WRITE` option, and it didn't fail. So, I'm guessing the option is not required. In fact, `make check` passes for me without errors.

**Comment by gc00** [comment]
This PR was migrated to #459 and pushed in.
